core/src/main/java/org/apache/accumulo/core/clientImpl/NamespaceMapping.java

```diff
@@ -0,0 +1,120 @@
+import com.google.gson.Gson;
+
+public class NamespaceMapping {
+  private static final Gson gson = new Gson();
+  private final ClientContext context;
+  private volatile SortedMap<NamespaceId,String> currentNamespaceMap = emptySortedMap();
+  private volatile SortedMap<String,NamespaceId> currentNamespaceReverseMap = emptySortedMap();
+  private volatile long lastMzxid;
+
+  public NamespaceMapping(ClientContext context) {
+    this.context = context;
+  }
+
+  public static byte[] initializeNamespaceMap() {
+    Map<String,String> map = Map.of(Namespace.DEFAULT.id().canonical(), Namespace.DEFAULT.name(),
+        Namespace.ACCUMULO.id().canonical(), Namespace.ACCUMULO.name());
+    return serialize(map);
+  }
+
+  public static void writeNamespaceToMap(ZooReaderWriter zoo, String zPath, NamespaceId namespaceId,
+      String namespaceName)
+      throws InterruptedException, KeeperException, AcceptableThriftTableOperationException {
+    // The built-in namespaces were already added during init or upgrade and can't be changed
+    if (!Namespace.DEFAULT.id().equals(namespaceId)
+        && !Namespace.ACCUMULO.id().equals(namespaceId)) {
+      zoo.mutateExisting(zPath, data -> {
+        var namespaces = deserialize(data);
+        // TODO throw exception if namespace name already exists in map?
+        namespaces.put(namespaceId.canonical(), namespaceName);
+        return serialize(namespaces);
+      });
+    }
+  }
+
+  public static byte[] serialize(Map<String,String> map) {
+    var sortedMap = ImmutableSortedMap.<String,String>naturalOrder().putAll(map).build();
+    String jsonData = gson.toJson(sortedMap);
+    return jsonData.getBytes(UTF_8);
+  }
+
+  public static Map<String,String> deserialize(byte[] data) {
+    if (data == null) {
+      throw new AssertionError("/namespaces node should not be null");
+    }
+    String jsonData = new String(data, UTF_8);
+    Type type = new TypeToken<Map<String,String>>() {}.getType();
+    return gson.fromJson(jsonData, type);
+  }
+
+  private synchronized void update() {
+    final ZooCache zc = context.getZooCache();
+    final String zPath = context.getZooKeeperRoot() + Constants.ZNAMESPACES;
+    final ZooCache.ZcStat stat = new ZooCache.ZcStat();
+
+    byte[] data = zc.get(zPath, stat);
+    if (stat.getMzxid() > lastMzxid) {
+      if (data == null) {
+        throw new AssertionError("/namespaces node should not be null");
+      } else {
+        Map<String,String> idToName = deserialize(data);
+        var converted = ImmutableSortedMap.<NamespaceId,String>naturalOrder();
+        var convertedReverse = ImmutableSortedMap.<String,NamespaceId>naturalOrder();
+        idToName.forEach((idString, name) -> {
+          var id = NamespaceId.of(idString);
+          converted.put(id, name);
+          convertedReverse.put(name, id);
+        });
+        currentNamespaceMap = converted.build();
+        currentNamespaceReverseMap = convertedReverse.build();
+      }
```

Comment:
Is it always expected that the desreialized map will contain the accumulo namepsace?  If so could check for that when deserializing and throw an exception if its not seen.